test/module/irohad/multi_sig_transactions/transport_test.cpp

```diff
@@ -87,17 +92,112 @@ TEST(TransportTest, SendAndReceive) {
   // we want to ensure that server side will call onNewState()
   // with same parameters as on the client side
   EXPECT_CALL(*notifications, onNewState(_, _))
-      .WillOnce(Invoke(
-          [&my_key, &cv, &state](const auto &from_key, auto const &target_state) {
-            EXPECT_EQ(my_key.publicKey(), from_key);
+      .WillOnce(Invoke([&my_key, &cv, &state](const auto &from_key,
+                                              auto const &target_state) {
+        EXPECT_EQ(my_key.publicKey(), from_key);
 
-            EXPECT_EQ(state, target_state);
-            cv.notify_one();
-          }));
+        EXPECT_EQ(state, target_state);
+        cv.notify_one();
+      }));
 
   transport->sendState(*peer, state);
   std::unique_lock<std::mutex> lock(mtx);
   cv.wait_for(lock, std::chrono::milliseconds(5000));
 
   server->Shutdown();
 }
+
+/**
+ * Checks that replayed transactions would not pass MST
+ * (receiving of already processed transactions would not cause new state
+ * generation)
+ * @given an instance of MstTransportGrpc
+ * @when exactly the same state reaches MST two times (in general, the state can
+ * be different but should contain exactly the same batch both times)
+ * @then for the first time the mock of tx_presence_cache says that the
+ * transactions of the batch are not found in cache, and mst produced and
+ * propagated new state that contains the batch. At the second time, the mock of
+ * tx_presence cache says that the transactions from the batch were previously
+ * rejected, so the state propagated via onNewState call would not contain the
+ * test batch.
+ */
+TEST(TransportTest, ReplayAttack) {
+  auto async_call_ = std::make_shared<
```

Comment:
A lot of initialization code duplication in two tests of this file. Maybe refactor it in a fixture?